theano/sandbox/cuda/dnn.py

```diff
@@ -2186,6 +2186,460 @@ def method(self):
+                                for bn_desc in self.bn_descs]
+
+    def infer_shape(self, node, shape):
+        if self.direction == 'inference':
+            return [shape[0]]
+        elif self.direction == 'forward':
+            return [shape[0], shape[1], shape[2], shape[3], shape[4]]
+        else:
+            return [shape[0], shape[2], shape[2]]
+
+    def _define_tensor4d_desc(self, name, id):
+        return """
+cudnnTensorDescriptor_t %(id)s_%(name)s;
+""" % dict(name=name, id=id)
+
+    def _init_tensor4d_desc(self, name, id, fail):
+        return """
+%(id)s_%(name)s = NULL;
+if ((err%(name)s = cudnnCreateTensorDescriptor(&%(id)s_%(name)s)) != CUDNN_STATUS_SUCCESS) {
+  PyErr_Format(PyExc_MemoryError, "could not allocate tensor descriptor "
+               ": %%s", cudnnGetErrorString(err%(name)s));
+  %(fail)s
+}
+""" % dict(name=name, id=id, fail=fail)
+
+    def _clean_tensor4d_desc(self, name, id):
+        return """
+if(%(id)s_%(name)s!= NULL)
+  cudnnDestroyTensorDescriptor(%(id)s_%(name)s);
+""" % dict(name=name, id=id)
+
+    def c_support_code_struct(self, node, name):
+        result = ''
+        for id in self.tensor_4d_descs:
+            result += self._define_tensor4d_desc(name, id)
+        return result
+
+    def c_init_code_struct(self, node, name, sub):
+        result = """
+cudnnStatus_t err%(name)s;
+""" % dict(name=name)
+
+        for id in self.tensor_4d_descs:
+            result += self._init_tensor4d_desc(name, id, sub['fail'])
+        return result
+
+    def c_cleanup_code_struct(self, node, name):
+        result = ''
+        for id in self.tensor_4d_descs:
+            result += self._clean_tensor4d_desc(name, id)
+        return result
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        if self.tensor_format == 'b01c':
+            tensor_format = 1
+        else:
+            tensor_format = 0
+
+        if self.mode == "spatial":
+            mode = "CUDNN_BATCHNORM_SPATIAL"
+        else:
+            mode = "CUDNN_BATCHNORM_PER_ACTIVATION"
+
+        # Setup configuration variables.
+        result = """
+cudnnStatus_t err%(name)s;
+cudnnTensorFormat_t format%(name)s = CUDNN_TENSOR_NCHW;
+if (%(tensor_format)d == 1)
+  format%(name)s = CUDNN_TENSOR_NHWC;
+
+cudnnBatchNormMode_t mode%(name)s = %(mode)s;
+
+double exponentialAverageFactor%(name)s = %(exp_avg_factor)f;
+
+double epsilon%(name)s = %(epsilon)e;
+""" % dict(name=name, tensor_format=tensor_format, mode=mode, exp_avg_factor=self.exp_avg_factor,
+           epsilon=self.epsilon)
+
+        # Validate the input and build the input variables.
+        for input_idx, bn_input in enumerate(inputs):
+            if self.direction == 'backward':
+                if input_idx == 0:
+                    input_name = 'bn_input'
+                elif input_idx == 1:
+                    input_name = 'bn_gout'
+                else:
+                    input_name = 'scale_bias_mean'
+            else:
```

Comment:
I think I understand now this loop over input and the second loop over output. Both of them are to create the cudnn tensor descriptors. But as Theano have more inputs + outputs then the number of descriptor, this is probably the cause of the problems you have.

I think you should remove both of those loop and make one over self.bn_desc. I would change self.bn_decs to be something like
bn_descs = [('name', input_index), ...]
this way, you know the good mapping between which input need a descriptor and only init those one.

As all descriptor are on tensor that are in the inputs of the node, you don't need to create some for the output of this node. This is because this node don't allocate output memory, but write on some of its inputs.
I think you should have